// proxima-autogen/src/proxima-sdk-vertex-client/index.js
"use strict";

const vertex = require("./proximaClientGenerator.js");
const { exec } = require("child_process");
const fs = require("fs-extra");
//IMPORTS
const {
  schemaTypescriptGenerator,
  ethTypescriptGenerator,
} = require("./typescriptGenerators.js");

async function generateEthereumTypescriptTemplates(app_config = {}) {
  let abiPath = "./abi";
  let typescriptOutputPath = "./DataAggregator/generated/contracts";
  let schemaFile = app_config.schema.file;
  let resp = await runEthereumCommand(abiPath, typescriptOutputPath);

  ethTypescriptGenerator.updateEthereumTypescriptTypes(
    schemaFile,
    typescriptOutputPath
  );
}

async function runCommandLineArgs(commandList) {
  //for each in commandLine, run the argument, check for errors, print args
  let commandLineText = commandList.join(" && "); //map to add " && " between values join
  //console.log(commandLineText);
  let resp = await new Promise((resolve, reject) => {
    exec(commandLineText, function (error, stdout, stderr) {
      console.log("stdout: " + stdout);
      console.log("stderr: " + stderr);
      if (error !== null) {
        console.log("exec error: " + error);
        // Reject if there is an error:
        return reject(error);
      }
      // Otherwise resolve the promise:
      resolve();
    });
  }, 40000);
  return resp;
}

async function runEthereumCommand(
  abiPath,
  typescriptOutputPath,
  targetVersion = "ethers-v5"
) {
  let installArg =
    "npm install --save-dev typechain @typechain/ethers-v5 @typechain/ethers-v4 @typechain/truffle-v5 @typechain/web3-v1 @typechain/truffle-v4 @typechain/hardhat";
  let genArg =
    "typechain --target $targetVersion --outDir '$typescriptOutputPath' '$abiPath/*.json'";
  genArg = genArg
    .replace("$abiPath", abiPath)
    .replace("$typescriptOutputPath", typescriptOutputPath)
    .replace("$targetVersion", targetVersion);
  let resp = await runCommandLineArgs([installArg, genArg]);
}

async function generateSchemaTypescriptTemplates(app_config = {}) {
  //console.log(app_config);
  let schemaFile = app_config.schema.file;
  let outputConfigGenFile = "./codegen.yml";
  //here
  let typescriptOutputPath = "./DataAggregator/generated/models";
  //typescriptOutputPath + "/models.ts"
  fs.ensureDirSync(typescriptOutputPath);
  schemaTypescriptGenerator.updateTypescriptGenConfiguration(
    schemaFile,
    typescriptOutputPath,
    outputConfigGenFile
  );

  let resp = await runSchemaCommand(schemaFile, typescriptOutputPath);
  //console.log(app_config);
  schemaTypescriptGenerator.updateSchemaTypescriptTypes(
    schemaFile,
    typescriptOutputPath
  );
}

async function runSchemaCommand(schemaFile, typescriptOutputPath) {
  let installArgs = [
    "npm install --save graphql",
    "npm install --save-dev @graphql-codegen/cli",
    "npm install --save-dev @graphql-codegen/typescript",
    "npm install --save-dev @graphql-codegen/typescript-operations",
    "npm install --save-dev @graphql-codegen/typescript-graphql-request",
  ];

  let resp = await runCommandLineArgs(installArgs);

  let genArg = "npx graphql-codegen --config ./codegen.yml";
  resp = await runCommandLineArgs([genArg]);
}

//utils
function generateProximaSDKVertexClient(config) {
  let vertexConfig = {};
  vertexConfig.name = config.name;
  vertexConfig.schemaFile = config.schema.file;
  vertex.generateVertexClient(vertexConfig);
}

module.exports = {
  generateProximaSDKVertexClient,
  generateEthereumTypescriptTemplates,
  generateSchemaTypescriptTemplates,
};
